pinot-common/src/main/java/org/apache/pinot/common/function/FunctionRegistry.java

```diff
@@ -94,83 +155,196 @@ private FunctionRegistry() {
+      }
+    } else if (sqlOperator instanceof PinotSqlScalarFunction) {
+      return ((PinotSqlScalarFunction) sqlOperator).getFunction();
+    }
+    return null;
   }
 
-  public static Set<String> getRegisteredCalciteFunctionNames() {
-    return FUNCTION_MAP.map().keySet();
+  public static NameMultimap<PinotFunction> getFunctionMap() {
+    return FUNCTION_MAP;
   }
 
-  /**
-   * Returns {@code true} if the given function name is registered, {@code false} otherwise.
-   */
-  public static boolean containsFunction(String functionName) {
-    return FUNCTION_INFO_MAP.containsKey(canonicalize(functionName));
+  public static NameMultimap<SqlOperator> getOperatorMap() {
+    return OPERATOR_MAP;
   }
 
-  /**
-   * Returns the {@link FunctionInfo} associated with the given function name and number of parameters, or {@code null}
-   * if there is no matching method. This method should be called after the FunctionRegistry is initialized and all
-   * methods are already registered.
-   */
-  @Nullable
-  public static FunctionInfo getFunctionInfo(String functionName, int numParameters) {
-    Map<Integer, FunctionInfo> functionInfoMap = FUNCTION_INFO_MAP.get(canonicalize(functionName));
-    if (functionInfoMap != null) {
-      FunctionInfo functionInfo = functionInfoMap.get(numParameters);
-      if (functionInfo != null) {
-        return functionInfo;
+  private static void registerFunction(Method method, Set<String> alias, boolean nullableParameters, boolean varArg) {
+    if (method.getAnnotation(Deprecated.class) == null) {
+      for (String name : alias) {
+        registerCalciteNamedFunctionMap(name, method, nullableParameters, varArg);
       }
-      return functionInfoMap.get(VAR_ARG_KEY);
     }
-    return null;
   }
 
-  private static String canonicalize(String functionName) {
-    return StringUtils.remove(functionName, '_').toLowerCase();
+  private static void registerFunction(Class<?> clazz, Set<String> alias, boolean nullableParameters, boolean varArg) {
+    if (clazz.getAnnotation(Deprecated.class) == null) {
+      for (String name : alias) {
+        registerCalciteNamedFunctionMap(name, clazz, nullableParameters, varArg);
+      }
+    }
+  }
+
+  private static void registerCalciteNamedFunctionMap(String name, Method method, boolean nullableParameters,
+      boolean varArg) {
+    FUNCTION_MAP.put(name, new PinotScalarFunction(name, method, nullableParameters, varArg));
+  }
+
+  private static void registerCalciteNamedFunctionMap(String name, Class<?> clazz, boolean nullableParameters,
+      boolean varArg) {
+    try {
+      SqlReturnTypeInference returnTypeInference =
+          (SqlReturnTypeInference) clazz.getField("RETURN_TYPE_INFERENCE").get(null);
+      SqlOperandTypeChecker operandTypeChecker =
+          (SqlOperandTypeChecker) clazz.getField("OPERAND_TYPE_CHECKER").get(null);
+      for (Method method : clazz.getMethods()) {
+        if (method.getName().equals("eval")) {
```

Comment:
Small improvement: Could we annotate the methods instead of forcing to use a special method name? I guess the difference is not huge and it may be a personal preference, but it can be easier to look for implementations by looking for usages of an annotation.

Reply:
Same with special fields like `RETURN_TYPE_INFERENCE` or `OPERAND_TYPE_CHECKER`. Why not just force these classes to implement a method that returns these values?

Reply:
both are great suggestions/improvements. there's no reason, i simply fell 
1. there's no need to annotate additionally to all callable methods if we already have the class annotated (might be verbose with lots of polymorphism)
2. for the special fields yeah we can use interface method stub. it is just didn't occur to me that we need different ones within the same annotated function class

Reply:
I see. I guess it won't be in the short term, but I think we strongly need to move from a system where `FunctionInvoker` does `_method.invoke(_instance, args)` to a system where the equivalent to `FunctionInvoker` always call `PinotFunction.eval(Object[])`.

`PinotFunction` would be defined as:

```java
public interface PinotFunction extends Function {
  SqlOperandTypeChecker getOperandTypeChecker();
  SqlReturnTypeInference getReturnTypeInference();
  Object eval(Object[] args);
}
```

```java
public abstract class SumPinotFunction implements PinotFunction {
    public SqlReturnTypeInference getReturnTypeInference() {
       return ReturnTypes.NULLABLE_SUM;
    }

   public Object eval(Object[] args) {
      if (args[0] == null) return args[1];
      if (args[1] == null) return args[0];
      return evalNotNull(args);
   }

   public abstract evalNotNull(Object[] args);

  @ScalarFunction
  public class SumIntPinotFunctionSupplier {
    public SqlOperandTypeChecker getOperandTypeChecker() {
       return // something that defines int args and arity 2
    }

    public Object evalNotNull(Object[] args) {
       int i1 = args[0];
       int i2 = args[1];

       return i1 + i1
    }
  }
  // Same with the other options
}
```